src/hotspot/share/gc/shenandoah/shenandoahCollectorPolicy.cpp

```diff
@@ -119,10 +116,32 @@ void ShenandoahCollectorPolicy::print_gc_stats(outputStream* out) const {
 
   size_t completed_gcs = _success_full_gcs + _success_degenerated_gcs + _success_concurrent_gcs;
   out->print_cr(SIZE_FORMAT_W(5) " Completed GCs", completed_gcs);
-  out->print_cr(SIZE_FORMAT_W(5) " Successful Concurrent GCs (%.2f%%)",  _success_concurrent_gcs, percent_of(_success_concurrent_gcs, completed_gcs));
-  out->print_cr("  " SIZE_FORMAT_W(5) " invoked explicitly (%.2f%%)",    _explicit_concurrent, percent_of(_explicit_concurrent, _success_concurrent_gcs));
-  out->print_cr("  " SIZE_FORMAT_W(5) " invoked implicitly (%.2f%%)",    _implicit_concurrent, percent_of(_implicit_concurrent, _success_concurrent_gcs));
-  out->print_cr("  " SIZE_FORMAT_W(5) " abbreviated (%.2f%%)",           _abbreviated_concurrent_gcs, percent_of(_abbreviated_concurrent_gcs, _success_concurrent_gcs));
+
+  size_t explicit_requests = 0;
+  size_t implicit_requests = 0;
+  for (int c = 0; c < GCCause::_last_gc_cause; c++) {
+    size_t cause_count = _collection_causes[c];
+    if (cause_count > 0) {
+      auto cause = (GCCause::Cause) c;
+      if (is_explicit_gc(cause)) {
+        explicit_requests += cause_count;
+      } else if (is_implicit_gc(cause)) {
+        implicit_requests += cause_count;
+      }
+      const char* desc = GCCause::to_string(cause);
+      out->print_cr("  " SIZE_FORMAT_W(5) " caused by %s (%.2f%%)", cause_count, desc, percent_of(cause_count, completed_gcs));
+    }
+  }
+
+  out->cr();
+  out->print_cr(SIZE_FORMAT_W(5) " Successful Concurrent GCs (%.2f%%)", _success_concurrent_gcs, percent_of(_success_concurrent_gcs, completed_gcs));
+  if (ExplicitGCInvokesConcurrent) {
```

Comment:
Instead of relying on flags here, should we explicitly (pun intended) record `_explicit_collection_causes` and `_implicit_collection_causes`?

Reply:
That is what used to happen in `ShenandoahControlThread`. It feels like duplicating a result that can be derived from recording the gc causes. The `ExplicitGCInvokesConcurrent` and `ShenandoahImplicitGCInvokesConcurrent` flags aren't manageable, so I don't expect them to be different when the report is generated. I suppose that could change in the future... I'd be more inclined to not tally them at all and just leave the report to show the individual causes without classifying them.